// @ngx-share/core/esm5/lib/share.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Inject, Injectable, Optional } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { CONFIG } from './share.models';
import { shareButtonsProp } from './share.prop';
import { mergeDeep } from './utils';
import * as i0 from "@angular/core";
import * as i1 from "./share.models";
var ShareService = /** @class */ (function () {
    function ShareService(config) {
        this.config = {
            prop: shareButtonsProp,
            options: {
                theme: 'default',
                include: [],
                exclude: [],
                size: 0,
                url: null,
                title: null,
                description: null,
                image: null,
                tags: null,
                twitterAccount: null,
                autoSetMeta: true,
                gaTracking: false,
                windowWidth: 800,
                windowHeight: 500,
                moreButtonIcon: 'ellipsis-h',
                lessButtonIcon: 'minus'
            }
        };
        this.config$ = new BehaviorSubject(this.config);
        this.setConfig(config);
    }
    Object.defineProperty(ShareService.prototype, "prop", {
        get: /**
         * @return {?}
         */
        function () {
            return this.config.prop;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ShareService.prototype, "twitterAccount", {
        get: /**
         * @return {?}
         */
        function () {
            return this.config.options.twitterAccount;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ShareService.prototype, "theme", {
        get: /**
         * @return {?}
         */
        function () {
            return this.config.options.theme;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ShareService.prototype, "windowSize", {
        get: /**
         * @return {?}
         */
        function () {
            return "width=" + this.config.options.windowWidth + ", height=" + this.config.options.windowHeight;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ShareService.prototype, "url", {
        get: /**
         * @return {?}
         */
        function () {
            return this.config.options.url;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ShareService.prototype, "title", {
        get: /**
         * @return {?}
         */
        function () {
            return this.config.options.title;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ShareService.prototype, "description", {
        get: /**
         * @return {?}
         */
        function () {
            return this.config.options.description;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ShareService.prototype, "image", {
        get: /**
         * @return {?}
         */
        function () {
            return this.config.options.image;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ShareService.prototype, "tags", {
        get: /**
         * @return {?}
         */
        function () {
            return this.config.options.tags;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ShareService.prototype, "autoSetMeta", {
        get: /**
         * @return {?}
         */
        function () {
            return this.config.options.autoSetMeta;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ShareService.prototype, "gaTracking", {
        get: /**
         * @return {?}
         */
        function () {
            return this.config.options.gaTracking;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ShareService.prototype, "size", {
        get: /**
         * @return {?}
         */
        function () {
            return this.config.options.size;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} config
     * @return {?}
     */
    ShareService.prototype.setConfig = /**
     * @param {?} config
     * @return {?}
     */
    function (config) {
        if (config) {
            this.config = mergeDeep(this.config, config);
            this.config$.next(this.config);
        }
    };
    /**
     * @param {?} name
     * @param {?} data
     * @return {?}
     */
    ShareService.prototype.addButton = /**
     * @param {?} name
     * @param {?} data
     * @return {?}
     */
    function (name, data) {
        var _a;
        /** @type {?} */
        var config = {
            prop: tslib_1.__assign({}, shareButtonsProp, (_a = {}, _a[name] = data, _a))
        };
        this.setConfig(config);
    };
    ShareService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    ShareService.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [CONFIG,] }] }
    ]; };
    /** @nocollapse */ ShareService.ngInjectableDef = i0.defineInjectable({ factory: function ShareService_Factory() { return new ShareService(i0.inject(i1.CONFIG, 8)); }, token: ShareService, providedIn: "root" });
    return ShareService;
}());
export { ShareService };
if (false) {
    /** @type {?} */
    ShareService.prototype.config;
    /** @type {?} */
    ShareService.prototype.config$;
}
