// @ngx-share/core/bundles/ngx-share-core.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('rxjs'), require('rxjs/operators'), require('@angular/common'), require('@angular/common/http'), require('@angular/platform-browser'), require('@angular/cdk/platform')) :
    typeof define === 'function' && define.amd ? define('@ngx-share/core', ['exports', '@angular/core', 'rxjs', 'rxjs/operators', '@angular/common', '@angular/common/http', '@angular/platform-browser', '@angular/cdk/platform'], factory) :
    (factory((global['ngx-share'] = global['ngx-share'] || {}, global['ngx-share'].core = {}),global.ng.core,global.rxjs,global.rxjs.operators,global.ng.common,global.ng.common.http,global.ng.platformBrowser,global.ng.cdk.platform));
}(this, (function (exports,i0,rxjs,operators,common,http,platformBrowser,platform) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
     */
    /** @type {?} */
    var CONFIG = new i0.InjectionToken('SHARE_BUTTONS_CONFIG');

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
     */
    /**
     * Simple object check.
     * @param {?} item
     * @return {?}
     */
    function isObject(item) {
        return (item && typeof item === 'object' && !Array.isArray(item));
    }
    /**
     * Deep merge two objects.
     * @param {?} target
     * @param {...?} sources
     * @return {?}
     */
    function mergeDeep(target) {
        var sources = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            sources[_i - 1] = arguments[_i];
        }
        var _a, _b;
        if (!sources.length) {
            return target;
        }
        /** @type {?} */
        var source = sources.shift();
        if (isObject(target) && isObject(source)) {
            for (var key in source) {
                if (isObject(source[key])) {
                    if (!target[key]) {
                        Object.assign(target, (_a = {}, _a[key] = {}, _a));
                    }
                    mergeDeep(target[key], source[key]);
                }
                else {
                    Object.assign(target, (_b = {}, _b[key] = source[key], _b));
                }
            }
        }
        return mergeDeep.apply(void 0, __spread([target], sources));
    }
    /**
     * Returns a readable number from share count
     * @param {?} num
     * @param {?} digits
     * @return {?}
     */
    function shareCountFormatter(num, digits) {
        /** @type {?} */
        var si = [
            { value: 1E9, symbol: 'B' },
            { value: 1E6, symbol: 'M' },
            { value: 1E3, symbol: 'K' }
        ];
        /** @type {?} */
        var rx = /\.0+$|(\.[0-9]*[1-9])0+$/;
        for (var i = 0; i < si.length; i++) {
            if (num >= si[i].value) {
                return (num / si[i].value).toFixed(digits).replace(rx, '$1') + si[i].symbol;
            }
        }
        return num.toFixed(digits).replace(rx, '$1');
    }
    /**
     * Copy text to clipboard
     * @param {?} ref
     * @return {?}
     */
    function copyToClipboard(ref) {
        return rxjs.of(ref.metaTags.url).pipe(operators.tap(function (url) {
            // Create a hidden TextArea element
            /** @type {?} */
            var textArea = ( /** @type {?} */(ref.document.createElement('textarea')));
            textArea.value = url;
            ref.document.body.appendChild(textArea);
            // highlight TextArea to copy the text
            if (ref.platform === 'ios') {
                /** @type {?} */
                var range = ref.document.createRange();
                range.selectNodeContents(textArea);
                /** @type {?} */
                var selection = ref.document.defaultView.getSelection();
                selection.removeAllRanges();
                selection.addRange(range);
                textArea.readOnly = true;
                textArea.setSelectionRange(0, 999999);
            }
            else {
                textArea.select();
            }
            ref.document.execCommand('copy');
            ref.document.body.removeChild(textArea);
        }));
    }
    /**
     * Returns a valid URL or falls back to current URL
     * @param {?} url
     * @param {?} fallbackUrl
     * @return {?}
     */
    function getValidUrl(url, fallbackUrl) {
        if (url) {
            /** @type {?} */
            var r = /(http|https):\/\/(\w+:{0,1}\w*@)?(\S+)(:[0-9]+)?(\/|\/([\w#!:.?+=&%@!\-\/]))?/;
            if (r.test(url))
                return url;
            console.warn("[ShareButtons]: Sharing link '" + url + "' is invalid!");
        }
        return fallbackUrl;
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
     */
    /**
     * Meta tags operator - Serialize meta tags into the sharer URL
     * @type {?}
     */
    var metaTagsOperators = [
        operators.map(function (ref) {
            // Social network sharer URL */
            /** @type {?} */
            var SharerURL = ref.prop.share[ref.platform];
            if (SharerURL) {
                // object contains supported meta tags
                /** @type {?} */
                var metaTags = ref.prop.share.metaTags;
                // object contains meta tags values */
                /** @type {?} */
                var metaTagsValues_1 = ref.metaTags;
                /** @type {?} */
                var link = '';
                // Set each meta tag with user value
                if (metaTags) {
                    link = Object.entries(metaTags).map(function (_a) {
                        var _b = __read(_a, 2), key = _b[0], metaTag = _b[1];
                        return metaTagsValues_1[key] ? metaTag + "=" + encodeURIComponent(metaTagsValues_1[key]) : '';
                    }).join('&');
                }
                return SharerURL + link;
            }
            return;
        })
    ];
    /**
     * Print button operator
     * @type {?}
     */
    var printOperators = [
        operators.map(function (ref) { return ref.document.defaultView.print(); })
    ];
    /**
     * Copy link to clipboard, used for copy button
     * @type {?}
     */
    var copyOperators = [
        operators.map(function (ref) {
            // Disable the button
            ref.renderer.setStyle(ref.el, 'pointer-events', 'none');
            ref.temp = { text: ref.prop.text, icon: ref.prop.icon };
            ref.metaTags.url = decodeURIComponent(ref.metaTags.url);
            return ref;
        }),
        operators.switchMap(function (ref) {
            return copyToClipboard(ref).pipe(operators.map(function () {
                ref.prop.text = ref.prop.successText;
                ref.prop.icon = ref.prop.successIcon;
                return ref;
            }), operators.catchError(function () {
                ref.prop.text = ref.prop.failText;
                ref.prop.icon = ref.prop.failIcon;
                return rxjs.of(ref);
            }));
        }),
        operators.map(function (ref) {
            ref.cd.markForCheck();
            return ref;
        }),
        operators.delay(2000),
        operators.map(function (ref) {
            // Enable the button
            ref.renderer.setStyle(ref.el, 'pointer-events', 'auto');
            // Reset copy button text and icon */
            ref.prop.text = ref.temp.text;
            ref.prop.icon = ref.temp.icon;
            ref.cd.markForCheck();
        })
    ];
    /**
     * Add the share URL to message body, used for WhatsApp and Email buttons
     * @type {?}
     */
    var urlInMessageOperators = [
        operators.map(function (ref) {
            /** @type {?} */
            var description = ref.metaTags.description;
            /** @type {?} */
            var url = ref.metaTags.url;
            /** @type {?} */
            var newRef = {
                metaTags: {
                    description: description ? description + "\r\n" + url : url
                }
            };
            return mergeDeep(ref, newRef);
        })
    ];
    /** @type {?} */
    var FacebookCountOperators = [
        operators.map(function (res) { return +res.share.share_count; })
    ];
    /** @type {?} */
    var PinterestCountOperators = [
        operators.map(function (text) { return JSON.parse(text.replace(/^receiveCount\((.*)\)/, '$1')); }),
        operators.map(function (res) { return +res.count; })
    ];
    /** @type {?} */
    var TumblrCountOperators = [
        operators.map(function (res) { return +res.response.note_count; })
    ];

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
     */
    /** @type {?} */
    var shareButtonsProp = {
        facebook: {
            type: 'facebook',
            text: 'Facebook',
            icon: ['fab', 'facebook-f'],
            color: '#4267B2',
            share: {
                desktop: 'https://www.facebook.com/sharer/sharer.php?',
                android: 'https://www.facebook.com/sharer/sharer.php?',
                ios: 'https://www.facebook.com/sharer/sharer.php?',
                operators: metaTagsOperators,
                metaTags: {
                    url: 'u'
                }
            },
            count: {
                request: 'http',
                url: 'https://graph.facebook.com?id=',
                operators: FacebookCountOperators
            }
        },
        twitter: {
            type: 'twitter',
            text: 'Twitter',
            icon: ['fab', 'twitter'],
            color: '#00acee',
            share: {
                desktop: 'https://twitter.com/intent/tweet?',
                android: 'https://twitter.com/intent/tweet?',
                ios: 'https://twitter.com/intent/tweet?',
                operators: metaTagsOperators,
                metaTags: {
                    url: 'url',
                    description: 'text',
                    tags: 'hashtags',
                    via: 'via'
                }
            }
        },
        google: {
            type: 'google',
            text: 'Google+',
            icon: ['fab', 'google-plus-g'],
            color: '#DB4437',
            share: {
                desktop: 'https://plus.google.com/share?',
                android: 'https://plus.google.com/share?',
                ios: 'https://plus.google.com/share?',
                metaTags: {
                    url: 'url',
                },
                operators: metaTagsOperators
            }
        },
        linkedin: {
            type: 'linkedin',
            text: 'LinkedIn',
            icon: ['fab', 'linkedin-in'],
            color: '#006fa6',
            share: {
                desktop: 'http://www.linkedin.com/shareArticle?',
                android: 'http://www.linkedin.com/shareArticle?',
                ios: 'http://www.linkedin.com/shareArticle?',
                operators: metaTagsOperators,
                metaTags: {
                    url: 'url',
                    title: 'title',
                    description: 'summary'
                },
            }
        },
        pinterest: {
            type: 'pinterest',
            text: 'Pinterest',
            icon: ['fab', 'pinterest-p'],
            color: '#BD091D',
            share: {
                desktop: 'https://in.pinterest.com/pin/create/button/?',
                android: 'https://in.pinterest.com/pin/create/button/?',
                ios: 'https://in.pinterest.com/pin/create/button/?',
                operators: metaTagsOperators,
                metaTags: {
                    url: 'url',
                    description: 'description',
                    image: 'media'
                }
            },
            count: {
                request: 'http',
                url: 'https://api.pinterest.com/v1/urls/count.json?url=',
                args: { responseType: 'text' },
                operators: PinterestCountOperators
            }
        },
        reddit: {
            type: 'reddit',
            text: 'Reddit',
            icon: ['fab', 'reddit-alien'],
            color: '#FF4006',
            share: {
                desktop: 'http://www.reddit.com/submit?',
                android: 'http://www.reddit.com/submit?',
                ios: 'http://www.reddit.com/submit?',
                operators: metaTagsOperators,
                metaTags: {
                    url: 'url',
                    title: 'title'
                },
            }
        },
        tumblr: {
            type: 'tumblr',
            text: 'Tumblr',
            icon: ['fab', 'tumblr'],
            color: '#36465D',
            share: {
                desktop: 'http://tumblr.com/widgets/share/tool?',
                android: 'http://tumblr.com/widgets/share/tool?',
                ios: 'http://tumblr.com/widgets/share/tool?',
                operators: metaTagsOperators,
                metaTags: {
                    url: 'canonicalUrl',
                    description: 'caption',
                    tags: 'tags'
                }
            },
            count: {
                request: 'jsonp',
                url: 'https://api.tumblr.com/v2/share/stats?url=',
                operators: TumblrCountOperators
            }
        },
        whatsapp: {
            type: 'whatsapp',
            text: 'WhatsApp',
            icon: ['fab', 'whatsapp'],
            color: '#25D366',
            share: {
                desktop: 'https://web.whatsapp.com/send?',
                android: 'whatsapp://send?',
                ios: 'whatsapp://send?',
                operators: __spread(urlInMessageOperators, metaTagsOperators),
                metaTags: {
                    description: 'text'
                }
            }
        },
        messenger: {
            type: 'messenger',
            text: 'Messenger',
            icon: ['fab', 'facebook-messenger'],
            color: '#0080FF',
            share: {
                android: 'fb-messenger://share/?',
                ios: 'fb-messenger://share/?',
                operators: metaTagsOperators,
                metaTags: {
                    url: 'link'
                }
            }
        },
        telegram: {
            type: 'telegram',
            text: 'Telegram',
            icon: ['fab', 'telegram-plane'],
            color: '#0088cc',
            share: {
                desktop: 'https://t.me/share/url?',
                android: 'https://t.me/share/url?',
                ios: 'https://t.me/share/url?',
                operators: metaTagsOperators,
                metaTags: {
                    url: 'url',
                    description: 'text'
                }
            }
        },
        vk: {
            type: 'vk',
            text: 'VKontakte',
            icon: ['fab', 'vk'],
            color: '#4C75A3',
            share: {
                desktop: 'http://vk.com/share.php?',
                android: 'http://vk.com/share.php?',
                ios: 'http://vk.com/share.php?',
                operators: metaTagsOperators,
                metaTags: {
                    url: 'url'
                }
            }
        },
        stumble: {
            type: 'stumble',
            text: 'Stumble',
            icon: ['fab', 'stumbleupon'],
            color: '#eb4924',
            share: {
                desktop: 'http://www.stumbleupon.com/submit?',
                android: 'http://www.stumbleupon.com/submit?',
                ios: 'http://www.stumbleupon.com/submit?',
                operators: metaTagsOperators,
                metaTags: {
                    url: 'url'
                }
            }
        },
        xing: {
            type: 'xing',
            text: 'Xing',
            icon: ['fab', 'xing'],
            color: '#006567',
            share: {
                desktop: 'https://www.xing.com/app/user?op=share&',
                android: 'https://www.xing.com/app/user?op=share&',
                ios: 'https://www.xing.com/app/user?op=share&',
                operators: metaTagsOperators,
                metaTags: {
                    url: 'url'
                }
            }
        },
        sms: {
            type: 'sms',
            text: 'SMS',
            icon: 'comment-alt',
            color: '#20c16c',
            share: {
                desktop: 'sms:?',
                android: 'sms:?',
                ios: 'sms:&',
                metaTags: {
                    description: 'body'
                },
                operators: __spread(urlInMessageOperators, metaTagsOperators)
            }
        },
        email: {
            type: 'email',
            text: 'Email',
            icon: 'envelope',
            color: '#FF961C',
            share: {
                desktop: 'mailto:?',
                android: 'mailto:?',
                ios: 'mailto:?',
                operators: __spread(urlInMessageOperators, metaTagsOperators),
                metaTags: {
                    title: 'subject',
                    description: 'body'
                }
            }
        },
        copy: {
            type: 'copy',
            text: 'Copy link',
            successText: 'Copied',
            successIcon: 'check',
            failText: 'Error',
            failIcon: 'exclamation',
            icon: 'link',
            color: '#607D8B',
            share: {
                operators: copyOperators
            }
        },
        print: {
            type: 'print',
            text: 'Print',
            icon: 'print',
            color: '#765AA2',
            share: {
                operators: printOperators
            }
        }
    };

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
     */
    var ShareService = /** @class */ (function () {
        function ShareService(config) {
            this.config = {
                prop: shareButtonsProp,
                options: {
                    theme: 'default',
                    include: [],
                    exclude: [],
                    size: 0,
                    url: null,
                    title: null,
                    description: null,
                    image: null,
                    tags: null,
                    twitterAccount: null,
                    autoSetMeta: true,
                    gaTracking: false,
                    windowWidth: 800,
                    windowHeight: 500,
                    moreButtonIcon: 'ellipsis-h',
                    lessButtonIcon: 'minus'
                }
            };
            this.config$ = new rxjs.BehaviorSubject(this.config);
            this.setConfig(config);
        }
        Object.defineProperty(ShareService.prototype, "prop", {
            get: /**
             * @return {?}
             */ function () {
                return this.config.prop;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ShareService.prototype, "twitterAccount", {
            get: /**
             * @return {?}
             */ function () {
                return this.config.options.twitterAccount;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ShareService.prototype, "theme", {
            get: /**
             * @return {?}
             */ function () {
                return this.config.options.theme;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ShareService.prototype, "windowSize", {
            get: /**
             * @return {?}
             */ function () {
                return "width=" + this.config.options.windowWidth + ", height=" + this.config.options.windowHeight;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ShareService.prototype, "url", {
            get: /**
             * @return {?}
             */ function () {
                return this.config.options.url;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ShareService.prototype, "title", {
            get: /**
             * @return {?}
             */ function () {
                return this.config.options.title;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ShareService.prototype, "description", {
            get: /**
             * @return {?}
             */ function () {
                return this.config.options.description;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ShareService.prototype, "image", {
            get: /**
             * @return {?}
             */ function () {
                return this.config.options.image;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ShareService.prototype, "tags", {
            get: /**
             * @return {?}
             */ function () {
                return this.config.options.tags;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ShareService.prototype, "autoSetMeta", {
            get: /**
             * @return {?}
             */ function () {
                return this.config.options.autoSetMeta;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ShareService.prototype, "gaTracking", {
            get: /**
             * @return {?}
             */ function () {
                return this.config.options.gaTracking;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ShareService.prototype, "size", {
            get: /**
             * @return {?}
             */ function () {
                return this.config.options.size;
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @param {?} config
         * @return {?}
         */
        ShareService.prototype.setConfig = /**
         * @param {?} config
         * @return {?}
         */
            function (config) {
                if (config) {
                    this.config = mergeDeep(this.config, config);
                    this.config$.next(this.config);
                }
            };
        /**
         * @param {?} name
         * @param {?} data
         * @return {?}
         */
        ShareService.prototype.addButton = /**
         * @param {?} name
         * @param {?} data
         * @return {?}
         */
            function (name, data) {
                var _a;
                /** @type {?} */
                var config = {
                    prop: __assign({}, shareButtonsProp, (_a = {}, _a[name] = data, _a))
                };
                this.setConfig(config);
            };
        ShareService.decorators = [
            { type: i0.Injectable, args: [{
                        providedIn: 'root'
                    },] }
        ];
        /** @nocollapse */
        ShareService.ctorParameters = function () {
            return [
                { type: undefined, decorators: [{ type: i0.Optional }, { type: i0.Inject, args: [CONFIG,] }] }
            ];
        };
        /** @nocollapse */ ShareService.ngInjectableDef = i0.defineInjectable({ factory: function ShareService_Factory() { return new ShareService(i0.inject(CONFIG, 8)); }, token: ShareService, providedIn: "root" });
        return ShareService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
     */
    var ShareDirective = /** @class */ (function () {
        function ShareDirective(meta, el, http$$1, platform$$1, renderer, cd, shareService, document) {
            this.meta = meta;
            this.el = el;
            this.http = http$$1;
            this.platform = platform$$1;
            this.renderer = renderer;
            this.cd = cd;
            this.shareService = shareService;
            this.document = document;
            /**
             * share window closed subscription (to unsubscribe if the button is destroyed before the share window closes)
             */
            this._shareWindowClosed = rxjs.Subscription.EMPTY;
            /**
             * Get share count
             */
            this.getCount = false;
            /**
             * Set meta tags from document head, useful when SEO is supported
             */
            this.autoSetMeta = this.shareService.autoSetMeta;
            /**
             * Meta tags inputs - initialized from the global options
             */
            this.url = this.shareService.url;
            this.title = this.shareService.title;
            this.description = this.shareService.description;
            this.image = this.shareService.image;
            this.tags = this.shareService.tags;
            /**
             * Stream that emits when share count is fetched
             */
            this.count = new i0.EventEmitter();
            /**
             * Stream that emits when share dialog is opened
             */
            this.opened = new i0.EventEmitter();
            /**
             * Stream that emits when share dialog is closed
             */
            this.closed = new i0.EventEmitter();
        }
        /** Share link on element click */
        /**
         * Share link on element click
         * @return {?}
         */
        ShareDirective.prototype.onClick = /**
         * Share link on element click
         * @return {?}
         */
            function () {
                var _this = this;
                var _a;
                if (this.platform.isBrowser) {
                    /** @type {?} */
                    var metaTags = this.autoSetMeta ? {
                        url: this.url,
                        title: this.title || this.getMetaTagContent('og:title'),
                        description: this.description || this.getMetaTagContent('og:description'),
                        image: this.image || this.getMetaTagContent('og:image'),
                        via: this.shareService.twitterAccount,
                        tags: this.tags,
                    } : {
                        url: this.url,
                        title: this.title,
                        description: this.description,
                        image: this.image,
                        tags: this.tags,
                        via: this.shareService.twitterAccount,
                    };
                    // Share the link
                    // @ts-ignore
                    (_a = rxjs.of({
                        el: this.el.nativeElement,
                        renderer: this.renderer,
                        prop: this.prop,
                        cd: this.cd,
                        document: this.document,
                        platform: this.getPlatform(),
                        metaTags: metaTags
                    })).pipe.apply(_a, __spread(this.prop.share.operators, [operators.tap(function (sharerURL) { return _this.share(sharerURL); }),
                        operators.take(1)])).subscribe();
                }
            };
        /**
         * @param {?} changes
         * @return {?}
         */
        ShareDirective.prototype.ngOnChanges = /**
         * @param {?} changes
         * @return {?}
         */
            function (changes) {
                var _this = this;
                if (this.platform.isBrowser) {
                    if (changes['shareButton'] && (changes['shareButton'].firstChange || changes['shareButton'].previousValue !== this.shareButton)) {
                        this.createShareButton(this.shareButton);
                    }
                    if (!this.url || (changes['url'] && changes['url'].previousValue !== this.url)) {
                        this.url = getValidUrl(this.autoSetMeta
                            ? this.url || this.getMetaTagContent('og:url')
                            : this.url, this.document.defaultView.location.href);
                        if (this.getCount && this.prop.count) {
                            this.shareCount(this.url).subscribe(function (count) { return _this.count.emit(count); });
                        }
                    }
                }
            };
        /**
         * @return {?}
         */
        ShareDirective.prototype.ngOnDestroy = /**
         * @return {?}
         */
            function () {
                this._shareWindowClosed.unsubscribe();
            };
        /**
         * Open sharing dialog
         * @param url - Share URL
         */
        /**
         * Open sharing dialog
         * @param {?} url - Share URL
         * @return {?}
         */
        ShareDirective.prototype.share = /**
         * Open sharing dialog
         * @param {?} url - Share URL
         * @return {?}
         */
            function (url) {
                var _this = this;
                if (url) {
                    // GA Tracking
                    if (this.shareService.gaTracking && (( /** @type {?} */(this.document.defaultView))).ga) {
                        (( /** @type {?} */(this.document.defaultView))).ga('send', 'social', this.prop.type, 'click', this.url);
                    }
                    // Open share pop up and activate its opened and closed events
                    this._shareWindowClosed = rxjs.of(this.document.defaultView.open(url, 'newwindow', this.shareService.windowSize)).pipe(operators.tap(function () { return _this.opened.emit(_this.prop.type); }), operators.switchMap(function (popUp) { return rxjs.interval(200).pipe(operators.takeWhile(function () { return !popUp.closed; })); }), operators.finalize(function () { return _this.closed.emit(_this.prop.type); })).subscribe();
                }
            };
        /**
         * @param {?} shareUrl
         * @return {?}
         */
        ShareDirective.prototype.shareCount = /**
         * @param {?} shareUrl
         * @return {?}
         */
            function (shareUrl) {
                var _a, _b;
                /** @type {?} */
                var options = this.prop.count;
                return options.request === 'jsonp'
                    ?
                        // @ts-ignore
                        (_a = this.http.jsonp(options.url + shareUrl, 'callback')).pipe.apply(_a, __spread(options.operators, [operators.catchError(function () { return rxjs.EMPTY; })])) :
                    // @ts-ignore
                    (_b = this.http.get(options.url + shareUrl, options.args)).pipe.apply(_b, __spread(options.operators, [operators.catchError(function () { return rxjs.EMPTY; })]));
            };
        /**
         * @param {?} buttonsName
         * @return {?}
         */
        ShareDirective.prototype.createShareButton = /**
         * @param {?} buttonsName
         * @return {?}
         */
            function (buttonsName) {
                /** @type {?} */
                var button = __assign({}, this.shareService.prop[buttonsName]);
                if (button) {
                    // Set share button properties
                    this.prop = button;
                    // Remove previous button class
                    this.renderer.removeClass(this.el.nativeElement, "sb-" + this._buttonClass);
                    // Add new button class
                    this.renderer.addClass(this.el.nativeElement, "sb-" + button.type);
                    // Set button css color variable
                    this.el.nativeElement.style.setProperty('--button-color', this.prop.color);
                    // Keep a copy of the class for future replacement
                    this._buttonClass = button.type;
                    // Set aria-label attribute
                    this.renderer.setAttribute(this.el.nativeElement, 'aria-label', button.ariaLabel || button.text);
                }
                else {
                    throw new Error("[ShareButtons]: The share button '" + buttonsName + "' does not exist!");
                }
            };
        /** Get meta tag content */
        /**
         * Get meta tag content
         * @param {?} key
         * @return {?}
         */
        ShareDirective.prototype.getMetaTagContent = /**
         * Get meta tag content
         * @param {?} key
         * @return {?}
         */
            function (key) {
                /** @type {?} */
                var metaUsingProperty = this.meta.getTag("property=\"" + key + "\"");
                if (metaUsingProperty)
                    return metaUsingProperty.getAttribute('content');
                /** @type {?} */
                var metaUsingName = this.meta.getTag("name=\"" + key + "\"");
                if (metaUsingName)
                    return metaUsingName.getAttribute('content');
            };
        /**
         * @return {?}
         */
        ShareDirective.prototype.getPlatform = /**
         * @return {?}
         */
            function () {
                if (this.platform.IOS)
                    return 'ois';
                if (this.platform.ANDROID)
                    return 'android';
                return 'desktop';
            };
        ShareDirective.decorators = [
            { type: i0.Directive, args: [{
                        selector: '[shareButton], [share-button]'
                    },] }
        ];
        /** @nocollapse */
        ShareDirective.ctorParameters = function () {
            return [
                { type: platformBrowser.Meta },
                { type: i0.ElementRef },
                { type: http.HttpClient },
                { type: platform.Platform },
                { type: i0.Renderer2 },
                { type: i0.ChangeDetectorRef },
                { type: ShareService },
                { type: undefined, decorators: [{ type: i0.Inject, args: [common.DOCUMENT,] }] }
            ];
        };
        ShareDirective.propDecorators = {
            shareButton: [{ type: i0.Input }],
            getCount: [{ type: i0.Input }],
            autoSetMeta: [{ type: i0.Input }],
            url: [{ type: i0.Input }],
            title: [{ type: i0.Input }],
            description: [{ type: i0.Input }],
            image: [{ type: i0.Input }],
            tags: [{ type: i0.Input }],
            count: [{ type: i0.Output }],
            opened: [{ type: i0.Output }],
            closed: [{ type: i0.Output }],
            onClick: [{ type: i0.HostListener, args: ['click',] }]
        };
        return ShareDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
     */
    var ShareCountPipe = /** @class */ (function () {
        function ShareCountPipe() {
        }
        /**
         * @param {?} num
         * @param {?=} digits
         * @return {?}
         */
        ShareCountPipe.prototype.transform = /**
         * @param {?} num
         * @param {?=} digits
         * @return {?}
         */
            function (num, digits) {
                return shareCountFormatter(num, digits);
            };
        ShareCountPipe.decorators = [
            { type: i0.Pipe, args: [{
                        name: 'shareCount'
                    },] }
        ];
        return ShareCountPipe;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
     */
    var ShareModule = /** @class */ (function () {
        function ShareModule() {
        }
        /**
         * @param {?=} config
         * @return {?}
         */
        ShareModule.forRoot = /**
         * @param {?=} config
         * @return {?}
         */
            function (config) {
                return {
                    ngModule: ShareModule,
                    providers: [
                        { provide: CONFIG, useValue: config }
                    ]
                };
            };
        ShareModule.decorators = [
            { type: i0.NgModule, args: [{
                        imports: [
                            platform.PlatformModule
                        ],
                        declarations: [
                            ShareDirective,
                            ShareCountPipe
                        ],
                        exports: [
                            ShareDirective,
                            ShareCountPipe
                        ]
                    },] }
        ];
        return ShareModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
     */

    exports.ShareModule = ShareModule;
    exports.ShareService = ShareService;
    exports.ShareDirective = ShareDirective;
    exports.metaTagsOperators = metaTagsOperators;
    exports.printOperators = printOperators;
    exports.copyOperators = copyOperators;
    exports.urlInMessageOperators = urlInMessageOperators;
    exports.FacebookCountOperators = FacebookCountOperators;
    exports.PinterestCountOperators = PinterestCountOperators;
    exports.TumblrCountOperators = TumblrCountOperators;
    exports.shareButtonsProp = shareButtonsProp;
    exports.CONFIG = CONFIG;
    exports.ShareCountPipe = ShareCountPipe;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
